src/rp2/plugin/accounting_method/hifo.py

```diff
@@ -0,0 +1,189 @@
+                index += 1
+
+        except StopIteration:
+            # End of acquired_lots
+            pass
+        self.__amount_list.sort()
+
+        
+    # AVL tree node keys have this format: <spot_price>_<internal_id>. The internal_id part is needed to disambiguate transactions
+    # that have the same spot_price. Internal_id is padded right in a string of fixed length (KEY_DISAMBIGUATOR_LENGTH).
+    def _get_avl_node_key(self, spot_price: RP2Decimal, internal_id: str) -> str:
+        return f"{spot_price}_{internal_id:0>{self.KEY_DISAMBIGUATOR_LENGTH}}"
+
+    # This function calls _get_avl_node_key with internal_id=MAX_KEY_DISAMBIGUATOR, so that the generated key is larger than any other key
+    # with the same timestamp.
+    def _get_avl_node_key_with_max_disambiguator(self, spot_price: RP2Decimal) -> str:
+        return self._get_avl_node_key(spot_price, self.MAX_KEY_DISAMBIGUATOR)
+
+    def get_next_taxable_event_and_amount(
+        self,
+        taxable_event: Optional[AbstractTransaction],
+        acquired_lot: Optional[InTransaction],
+        taxable_event_amount: RP2Decimal,
+        acquired_lot_amount: RP2Decimal,
+    ) -> TaxableEventAndAcquiredLot:
+        new_acquired_lot: Optional[InTransaction] = acquired_lot
+        new_acquired_lot_amount: RP2Decimal = acquired_lot_amount - taxable_event_amount if acquired_lot is not None else ZERO
+
+        try:
+            new_taxable_event: AbstractTransaction = next(self.__taxable_event_iterator)
+        except StopIteration:
+            raise TaxableEventsExhaustedException() from None
+        new_taxable_event_amount: RP2Decimal = new_taxable_event.crypto_balance_change
+
+        # If the new taxable event has different year than the acquired lot (and it's not earn-typed), also get a new acquired lot from the new year
+        if taxable_event and taxable_event.timestamp < new_taxable_event.timestamp:
+            if acquired_lot:
+                # Cache old-year acquired_lot amount
+                self.__acquired_lot_2_partial_amount[acquired_lot] = new_acquired_lot_amount
+            (_, new_acquired_lot, _, new_acquired_lot_amount) = self.get_acquired_lot_for_taxable_event(
+                new_taxable_event, acquired_lot, new_taxable_event_amount, new_acquired_lot_amount
+            )
+
+        return TaxableEventAndAcquiredLot(
+            taxable_event=new_taxable_event,
+            acquired_lot=new_acquired_lot,
+            taxable_event_amount=new_taxable_event_amount,
+            acquired_lot_amount=new_acquired_lot_amount,
+        )
+
+    def get_acquired_lot_for_taxable_event(
+        self, taxable_event: AbstractTransaction, acquired_lot: Optional[InTransaction], taxable_event_amount: RP2Decimal, acquired_lot_amount: RP2Decimal
+    ) -> TaxableEventAndAcquiredLot:
+        new_taxable_event_amount: RP2Decimal = taxable_event_amount - acquired_lot_amount
+        index_list = len(self.__amount_list)-1
```

Comment:
I was wondering why acquired_lot_list and acquired_lot_avl aren't bucketized by year to increase performance (similar to what LIFO does), but then I realized that in HIFO there are two constraints: time (can't sell lots that weren't bought yet) and spot_price (high to low). So this is more complicated than LIFO, which only has one constraint (time). 

So if we bucketized data, we would need to repeat all elements of year 2017 in the list and tree of year 2018, all the elements of 2018 in 2019, etc. This would make memory consumption roughly O(n^2). 

As it is, time complexity is O(n^2), which may be prohibitive for large numbers of transactions. Likely there are other optimizations we can do to improve this, but we can think about that later.

Anyway, this is fine for the first implementation, but maybe add a comment mentioning the above considerations as a reminder for future improvement.

Reply:
Something like:
```
# This while loop makes the algorithm's complexiti O(n^2), where n is the number of acquired lots. Non-trivial
# optimizations are possible using different data structures (and likely with some space/time tradeoff) 
```